feat: Serial dispatch (lazy action expansion)

cormacc opened the pull request:
This is a draft implementation of serial dispatch, per recent discussion in the nexus channel on clojurians:
https://clojurians.slack.com/archives/C09440Y2PK5/p1763568245754909?thread_ts=1762369001.162089&cid=C09440Y2PK5

# Why
The default nexus dispatch implementation expands actions eagerly and executes all effects together at the end. This allows for batching of effects, but at the cost of some trade-offs. Specifically:
- state effects expanded from actions early in a series are not executed prior to expanding subsequent actions -- i.e. actions may be interpolated and evaluated
with old state
- batching can interfere with execution order

Both of these make the system (subjectively) harder to reason about.

This changeset implements an alternative dispatch strategy that expands actions lazily / executes effects eagerly. During dispatch....
1. If the next dispatch item is an effect, interpolate only that item and execute it immediately
2. If not, expand the next item until the first item in the expanded vector is an effect
3. Goto 1

# Status
This is a draft / proposal -- I'll refine it a bit (and add some unit tests) if it's considered useful enough to others to take further.

It supports nexus interceptors -- the dataspex action log continues to function (well, mostly -- there are a couple of small gaps still to address in result accumulation)

It does not support batched effects. For UI use cases, the same result (preventing multiple renders during a single call to dispatch) can readily be achieved by adding a nexus interceptor that suspends rendering for the duration of dispatch.

In user-land, this would look something like...
``` clojure
;;=== Rendering
(defonce !render-status (atom {:dispatch-depth 0
                               :deferred 0}))

(defn render-if-idle [render-fn state]
  (if (= 0  (:dispatch-depth @!render-status))
    (render-fn state)
    (swap! !render-status update :deferred inc)))

(defn add-watches! [{:keys [!conn !store]} render-fn]
  (let [render-if-idle! (partial render-if-idle render-fn)]

    ;;... state atom
    (add-watch !store :on-change (fn [_ _ old-state new-state]
                                   (when (not= old-state new-state)
                                     (render-if-idle! (assoc new-state :db @!conn :now (tick/inst))))))
    ;;... datascript db
    (ds/listen! !conn ::render
                (fn [{:keys [db-after]}]
                  (render-if-idle! (assoc @!store :db  db-after :now (tick/inst)))))))

(defn get-deferred-render-interceptor [{:nexus/keys [system->state]} sys render-fn]
  {:id ::render
   :before-dispatch (fn [ctx]
                      (swap! !render-status update :dispatch-depth inc)
                      ctx)
   :after-dispatch (fn [ctx]
                     (swap! !render-status update :dispatch-depth dec)
                     (let [{:keys [dispatch-depth deferred]} @!render-status]
                       (when (and (= 0 dispatch-depth) (> deferred 0))
                         (swap! !render-status assoc :deferred 0)
                         (render-fn (system->state sys))))
                     ctx)})

(defn append-interceptor [nexus-map interceptor]
  (update nexus-map :nexus/interceptors conjv interceptor))

(defn set-dispatch [nexus-map system render-fn]
  (add-watches! system render-fn)
  (let [deferred-render-interceptor (get-deferred-render-interceptor nexus-map system render-fn)
        nexus-map* (append-interceptor nexus-map deferred-render-interceptor)]
    (r/set-dispatch!
     (fn [dispatch-data actions]
       (nexus.serial/dispatch nexus-map* system (assoc dispatch-data :now (tick/inst)) actions)))))
```

Ro0r committed:
feat: Serial dispatch (lazy action expansion)

The default nexus dispatch implementation expands actions eagerly and
executes all effects together at the end. This allows for batching of
effects, but at the cost of some trade-offs. Specifically:
- state effects expanded from actions early in a series are not executed
  prior to expanding subsequent actions -- i.e. actions may be interpolated
and evaluated
with old state
- batching can interfere with execution order

Both of these make the system (subjectively) harder to reason about.

This changeset implements an alternative dispatch strategy that expands
actions lazily / executes effects eagerly. During dispatch....
1. If the next dispatch item is an effect, interpolate only that item and execute it immediately
2. If not, expand the next item until the first item in the expanded vector is an effect
3. Goto 1

It supports nexus interceptors -- the dataspex action log continues to
function (well, mostly -- there are a couple of small gaps still to
address in result accumulation)

It does not support batched effects. For UI use cases, the same result
(preventing multiple renders during a single call to dispatch) can
readily be achieved by adding a nexus interceptor that suspends
rendering for the duration of dispatch (doc example to follow).

cormacc committed:
fix: Don't export the enable-dataspex function by default

Runtime enabling of dataspex is a decision that belongs in userland,
not the library.

Ro0r committed:
test: Added unit tests to verify expansion and execution order

Tests to verify eager expansion for default dispatch, and lazy expansion
for serial dispatch. Uses an interceptor in the nexus.test-helper
namespace.

cjohansen commented on src/nexus/serial.cljc:
```diff
@@ -0,0 +1,109 @@
+      ctx*
+      (recur nexus-map (expand-first nexus-map ctx*)))))
+
+(defn dispatch-serial-base [nexus-map {:keys [system dispatch-data] :as ctx}]
```
Perhaps `dispatch-serial-base`  could be `dispatch-serially` instead? It isn't clear to me what "base" is in this context.

cjohansen commented on src/nexus/serial.cljc:
```diff
@@ -0,0 +1,109 @@
+          (update :errors conjv {:phase :execute-effect
+                                 :effect-k effect-id
+                                 :err (ex-info "No such effect" {:available-effects (keys (:nexus/effects nexus-map))})})))))
+
```
More naming/style nitpicks, here's a suggestion with less let bindings and some naming changes:

```clj
(defn execute-first [{:nexus/keys [interceptors] :as nexus}
                     {:keys [dispatch-data actions] :as ctx}]
  (let [[effect-k :as effect] (first actions)]
    (if-let [effect-interceptor (get-effect-interceptor nexus effect)]
      (let [{:keys [results errors]}
            (nexus/run-interceptors
             (assoc ctx :effect (nexus/interpolate-1 nexus dispatch-data effect))
             (conjv interceptors effect-interceptor)
             [:before-effect :after-effect :effect])]
        (cond-> (assoc ctx :actions (rest actions))
          (seq results) (update :results conjv results)
          (seq errors) (update :errors conjv errors)))
      ;;Effect not found
      (-> ctx
          (assoc :actions (rest actions))
          (update :errors conjv
                  {:phase :execute-effect
                   :effect-k effect-k
                   :err (ex-info "No such effect"
                                 {:available-effects (keys (:nexus/effects nexus))})})))))
```

cjohansen reviewed (commented):
This looks very good! It's interesting that it can be implemented entirely with an interceptor. That makes me think that this feature should probably be offered as an interceptor, not a separate entry-point. This way you could use this strategy by doing something like `(nexus.serial/add-interceptor nexus)` or `(nexus.serial/install nexus)` or some such.

I imagine that this also breaks some of the expectations in the Dataspex action log. Did you have some ideas for how to handle that? I haven't thought a lot about it, but one idea is that maybe there could be a separate action-log implementation for this strategy, if that would make visualization easier/better aligned with what actually happens.

cormacc commented:
Thanks for the review Christian. Will work through this later today / this evening hopefully -- bogged down in some testing at the moment.

cormacc commented on src/nexus/serial.cljc:
```diff
@@ -0,0 +1,109 @@
+    @!dataspex-interceptor (update :nexus/interceptors conjv @!dataspex-interceptor)))
+
+;;=== Serial dispatch
+(defn get-effect-interceptor [nexus-map effect]
```
Done

cormacc committed:
refactor: Addressed code style feedback from @cjohansen

This changeset updates naming and code style to better align with nexus
codebase conventions, per initial feedback.

Ro0r committed:
test: Added tests for interpolation order

... although these probably don't add value over the existing dispatch
order tests

cormacc commented:
I saw that discussion. That should work too, but I have a couple of concerns....

1. It's making state a special case of effect. You'd presumably need to add a `:nexus/state->system` key to the map to re-integrate the state snapshot at the end of dispatch? And would you eliminate state effects then entirely, or consider how to reconcile any state effects generated during action expansion with mutations.
2. If your state snapshot contains data from multiple sources (I recall from a conversation at reclojure that at least one of your applications does, and my app's system->state function derefs an app state atom and a datascript DB) you may have some complex re-integration logic.

All in all it sounds like maybe a net gain in complexity / distributed workaround....

If the problem statement is (well, my problem statement is :) )
1. I'd like effects to be applied immediately in dispatched order to make the system as easy to reason about as possible.
2. I don't want updates to state to trigger unnecessary re-rendering

The first of these sounds like something that could be solved in nexus.

Quoting your own words out of context :) "I don't want nexus to know about rendering", the second of these should be solved in userland. And changing the context during dispatch as suggested (and to a lesser extent effect batching -- though that probably has some applications in effects using the network too) feels like pretending that nexus doesn't know about rendering, while actually implementing some workarounds under the table.

Though take all that with lots of salt -- I've gone a certain distance down a particular path so am definitely biased :)

Ro0r committed:
fix: Ensure async dispatch runs interceptors

Previously, the dispatch fn associated with dispatch-data for async
callbacks was just running the dispatch interceptor -- not any
additional interceptors added in userland (e.g. nexus action log)

Now the nested dispatch function runs all interceptors defined for the
before-dispatch and after-dispatch phases.

Should give a more useful action log. If you're adding a suspended
render interceptor in userland, this makes life a little more
complicated, as now you need to allow for nested/async dispatch there.
But just means incrementing and decrementing a nested dispatch level
counter rather than setting and clearing a semaphore...

Ro0r committed:
test(WIP): Reworking nexus core tests for serial dispatch

About halfway done

Commit:
Avoid error when an expanded action returns nil

In this case, it would fail the `nexus/actions?` check (because
`(sequential? nil)` is false), and so produce an error
"ACTION should expand to a collection of actions". The error is
harmless, but unnecessarily noisy.

Also removes a redundant fn.

cjohansen committed:
Nitpick some formatting

cjohansen committed:
Fix failing test

Expanding actions does not produce both actions and effects. expand-first does
one pass of expansion, `expand-lazily` figures out what is an action and what is
an effect.

cjohansen committed:
Move generally useful test helper to the helper namespace

cjohansen committed:
Enable and pass all tests in the serial test suite

cjohansen committed:
Rename nexus.serial to nexus.serial.dispatch

This enables a nice parallel layout for strategies:

- nexus.serial.dispatch and nexus.serial.registry
- nexus.batch.dispatch and nexus.batch.registry

cjohansen committed:
Add nexus.serial.dispatch

cjohansen committed:
Move log-interceptor to shared test-helper

cjohansen committed:
Move order test tooling to helper and split serial/batch cases

cjohansen committed:
Make sure batch and serial dispatch return the same value

cjohansen committed:
Add more batch-tests to the serial impl to increase parity

cjohansen committed:
Add test for the serial registry

Verify that actions can use state resulting from previous effects in the same
dispatch.

cormacc commented:
@cjohansen -- updated the deferred render interceptor example in the PR description. The change was necessary to allow for asynchronous dispatch, per commit 4ccba64ae78edcd38c46166055aa60fd849f43bf.
It's ugly, but seems to work.

cormacc commented on src/nexus/serial/registry.cljc:
```diff
@@ -0,0 +1,31 @@
+(ns nexus.serial.registry
```
Why not just use a single registry for both dispatch strategies? Is this to support some testing scenarios?

cjohansen replied:
I still haven't decided where I want to go with this, and took a step back to lull it over. The reason for this design choice was to be able to dead code eliminate the strategy you don't use. I want to sketch out an alternative tomorrow, and see where to take it.

cormacc committed:
fix: Preserve :queue and :stack context keys in serial dispatch interceptor

cormacc commented:
@cjohansen I've pushed an additional commit fixing an issue introduced by a984089ee500e7722edc3a8c796b5ef4b47e5538 -- it had broken the interceptor chain.

cormacc committed:
feat: Allow :before-dispatch interceptor to override dispatch on context

Previously we were binding the dispatch function for use in
effect callbacks in the serial dispatch :before-dispatch interceptor
itself. This meant that other interceptors couldn't replace it --
which they might want to do in certain testing scenarios.

This change binds the dispatch function to context in the call to
dispatch itself, before running interceptors, which should remove the
limitation.

See this discussion on the #nexus channel on Clojurians slack for one
scenario in which this might be useful:
https://clojurians.slack.com/archives/C09440Y2PK5/p1772116834092659

cormacc committed:
fix(test): Dissoc interceptor :queue and :stack prior to return

The result returned by serial dispatch retained the :queue and :stack
keys used during interceptor execution. Now dissocing them for
consistency with value returned by batch dispatch.

cormacc committed:
fix: Retain :dispatch on context after serial dispatch interceptor

Other :after-dispatch interceptors might need it after the serial
dispatch interceptor has run.

cjohansen commented:
I want to thank you @cormacc for all your efforts here! Since we decided to fix Nexus rather than add an alternate route, I've taken your ideas and reapplied them here: https://github.com/cjohansen/nexus/tree/instant-execute

I will look through this PR again for details that should be represented in the new implementation. I would love some feedback and/or testing from your end as well, to help me ensure it does everything right.

cormacc commented:
Thanks yourself Christian (and Magnar) for all the wonderful tools - I'm benefiting and learning hugely from them, and the distilled thought and experience they embody. Will poke at the branch early this week - started a refactor late Friday that I need to close out as first order of business tomorrow.

cormacc commented:
The changes over in the new branch LGTM. Will update my dev app to use that rather than my own fork tomorrow and poke at it some more.

cormacc commented:
Sorry @cjohansen -- I started a response here early in the week but mustn't have hit send.
An initial attempt to get up and running went off the tracks -- the (correct) decision in the new branch not to include a state snapshot in effect context any more exposed a good number of effects with pre-serial use of the state snapshot. Most of these were for relatively static state (auth state etc.), but still an anti-pattern.

Anyway I refactored them in bulk and broke everything, then had to abandon it and get back to work (if I were smarter I'd just have temporarily added a before-effect interceptor to inject a fresh state snapshot and corrected them one-by-one).

I switched back to my fork again and ended up eliminating this pattern in our own codebase while debugging a race condition with some fairly complex webserial comms.

This evening I decided to drop your new branch into our app again, but no instant gratification. I'll dig into it a bit next week to figure out just why it ain't working as a drop-in replacement. Weekend is looking busy with family commitments.

cjohansen commented:
Sorry to hear that @cormacc! You should've warned me before making lots of changes, removing the state from the effect handler context was not an intended change. I honestly wasn't aware that it was sent to effect handlers - it's not part of the intended contract. But if that's the way it is, there's no need to change it. I don't want the next release to require a bunch of work to upgrade - it should be as close to drop-in as possible given the wanted changes.

Let me know if you figure out what the problem is, and let's see if we can make changes in Nexus before you change your code to work around it 😊

cormacc commented:
I have a vague memory of spotting the state snapshot on dispatch-data while logging (keys dispatch-data) debugging something a while back, and thinking 'I wonder why that's there? Maybe deref is expensive and I should use it...'. - I'm still very much an enthusiastic amateur when it comes to Clojure. Reworking some effects last week to eliminate its use entirely and ensuring they explicitly deref !store at the point of use (and do so again in promise callbacks) has definitely tightened things up.

To give a concrete example -- our app reads data from a serial port in a go loop, which dispatches an action when new bytes are received. We had been seeing intermittent failures during firmware update that were hard to debug (occurring in one page write out of > 300). An audit of effects to ensure we expliticly deref store, and do so again in a chained .then handler made this intermittent failure occur consistently, and then very easy to fix. Before that, I'd been beginning to wonder if maybe the webserial API could episodically miss a byte, but the problem was (as usual) implementation error.

Short version, I'd suggest leaving the state snapshot off dispatch-data and maybe adding something to the docs stating that if you want to retain the old behaviour for legacy code you can do so easily using a before-effect interceptor?

Honestly, I'm happy to work through the issues -- swapping between things that should be interchangeable help flush out things in our code that may be working by happy accident and drives useful housekeeping. There are quite a few moving parts in our dev app -- will hopefully get some time next week to strip layers back and build them back up again incrementally and see just where the issue is. The straight drop-in Thursday resulted in a frozen browser tab and no dataspex, so debugging opportunities were limited (though I can probably do a bit of digging at the REPL when I'm back at the desk).

cormacc commented:
Finally got back to this today. My locking issues seem to be associated with nested dispatch, for whatever reason. I dispatch a number of effects on app start to initialise various subsystems, and dispatching any that use nested dispatch in an async callback cause the app to lock up -- hard freeze, can't refresh the browser tab - need to close it.

The issue doesn't appear to be with the bound dispatch function itself (on naive analysis) -- I have function defined for use from the repl that looks like this:
``` clojure
(defn dispatch! [args]
  #?(:cljs
     (replicant/*dispatch* {:replicant/trigger :replicant.trigger/external} args))
  ;;Return nil to avoid clogging up the repl
  nil)
```
And using that from the async callback instead of the bound function results in the same hard freeze (it continues to work fine from the repl with your new branch as it did with my fork).

So maybe the issue is some combination of async/nested dispatch and a subtle difference in interceptor sequencing...

I'll poke at it a bit more when I get a chance, and try to build up a minimal repro -- what I've been doing so far has been stripping back and building up layers of our main dev app, which has been instructive for me, but not so good for illustration :)

cormacc commented:
Aha! I have it. Well, I have to give eca/claude the credit for this one....

So in nexus.core, line 78....

``` clojure
          (empty? actions) ctx
```

Should be 
``` clojure
          (empty? actions) (select-keys ctx [:errors :trace])
```

As is, when an action handler returns no actions, the returned context contains actions from the parent dispatch loop that have already been expanded, causing an infinite loop.

So my earlier statement that it's definitely nested dispatch was definitely incorrect :)  
It just so happens that a couple of the my nested dispatch calls on app init will expand to nil/an empty vector if we haven't authenticated yet. Anyway, all looking good now after cloning your new branch and making the correction locally. Will give it some more extensive testing Monday.

cjohansen commented:
Fantastic investigation! Thanks so much 🙏 I added a test and fixed it: https://github.com/cjohansen/nexus/commit/a131756ec789f6ef54c7ec0d430163e71cf12fab

cormacc commented:
I've been poking a bit more at my residual issues. It appears that some actions / effects are getting executed out of order, when (next %) returns nil, we end

In nexus.core, line 196, it looks like

``` clojure
(update :actions #(into (next %) (:actions res)))

```
should be
``` clojure
(update :actions #(into (vec (:actions res)) (next %)))
```

cormacc commented:
Looks like my app is working now with the new branch. I'm going to shut down before I break something else :)

cjohansen commented:
Lovely, thanks again! https://github.com/cjohansen/nexus/commit/650edab060b78008e8d7c465691e9d4b6c231adb

cormacc commented:
Thanks yourself Christian -- all still looking good my end -- raised an MR yesterday evening replacing our fork with the latest commit on the new branch. Will report back to you if we run into any more residual issues.
